src/librustc_typeck/check/pat.rs

```diff
@@ -1146,7 +1181,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else {
             (tcx.types.err, tcx.types.err)
         };
-        self.check_pat(&inner, inner_ty, def_bm, discrim_span);
+        self.check_pat(&inner, inner_ty, def_bm, discrim_span, None);
```

Comment:
I'm not yet sold on the use of `None` and `Some(pat)`. For example, I don't see why e.g. `let box bar` or `let &bar` cannot also receive the new enhanced diagnostics. In fact, in what circumstance is `Some(pat)` unambiguously wrong?

On the side, I've been thinking that we could pass a trace to the top level pattern (all the way to `check_top_pat`) + the expected type via a stack-based linked list (which should incur no allocation and thus negligible cost).

Reply:
That sounds reasonable but I wanted to tackle the struct field case first and see into expanding the uses if parent_oat afterwards.

Reply:
I would prefer to do it now because I think the in-between situation complicates the code, and I suspect not really for a good reason, and also it should be easy to just remove the `Option<_>` layer and unconditionally pass the parent. A few tests might need to be added but it might also already be exercised.